src/idl_gen_csharp.cpp

```diff
@@ -1745,93 +1763,91 @@ class CSharpGenerator : public BaseGenerator {
-        auto tmp_name_vec = name_vec;
-        tmp_name_vec.push_back(field.name);
-        auto tmp_array_length_vec = array_length_vec;
-        if (is_array) {
-          tmp_array_length_vec.push_back(field_type.fixed_length);
+      FieldArrayLength tmpArrayLength = {
+        field.name,
+        field_type.fixed_length,
+      };
+      arrayLengths.push_back(tmpArrayLength);
+      if (field_type.struct_def != nullptr) {
+        GenStructPackDecl_ObjectAPI(*field_type.struct_def, code_ptr,
+                                    arrayLengths);
+      } else {
+        std::vector<FieldArrayLength> arrayOnlyLengths;
+        for (size_t i = 0; i < arrayLengths.size(); ++i) {
+          if (arrayLengths[i].length > 0) {
+            arrayOnlyLengths.push_back(arrayLengths[i]);
+          }
         }
         std::string name;
-        for (size_t tmp_name_index = 0; tmp_name_index < tmp_name_vec.size();
-             ++tmp_name_index) {
-          name += "_" + tmp_name_vec[tmp_name_index];
-        }
-        code += "    var " + name + " = new " + GenTypeBasic(field_type) + "[";
-        code += NumToString(tmp_array_length_vec[0]);
-        for (size_t i = 1; i < tmp_array_length_vec.size(); ++i) {
-          auto array_length = tmp_array_length_vec[i];
-          code += "," + NumToString(array_length);
-        }
-        code += "];\n";
-        code += "    ";
-        // initialize array
-        for (size_t i = 0; i < tmp_array_length_vec.size(); ++i) {
-          auto array_length = tmp_array_length_vec[i];
-          auto idx = "idx" + NumToString(i);
-          code += "for (var " + idx + " = 0; " + idx + " < " +
-                  NumToString(array_length) + "; ++" + idx + ") {";
-        }
-        code += name + "[idx0";
-        for (size_t i = 1; i < tmp_array_length_vec.size(); ++i) {
-          auto idx = "idx" + NumToString(i);
-          code += "," + idx;
+        for (size_t i = 0; i < arrayLengths.size(); ++i) {
+          name += "_" + arrayLengths[i].name;
         }
-        code += "] = _o";
-        for (size_t i = 0; i < tmp_array_length_vec.size(); ++i) {
-          auto idx = "idx" + NumToString(i);
-          code += "." + MakeCamel(tmp_name_vec[i]) + "[" + idx + "]";
-        }
-        if (!is_array) { code += "." + MakeCamel(field.name); }
-        code += ";";
-        for (size_t i = 0; i < tmp_array_length_vec.size(); ++i) {
-          code += "}";
+        code += "    var " + name + " = ";
+        if (arrayOnlyLengths.size() > 0) {
+          code += "new " + GenTypeBasic(field_type) + "[";
+          for (size_t i = 0; i < arrayOnlyLengths.size(); ++i) {
+            if (i != 0) { code += ","; }
+            code += NumToString(arrayOnlyLengths[i].length);
+          }
+          code += "];\n";
+          code += "    ";
+          // initialize array
+          for (size_t i = 0; i < arrayOnlyLengths.size(); ++i) {
+            auto idx = "idx" + NumToString(i);
```

Comment:
why are we copying things into a new array here? I'll need to see what code this generates. I would expect this change to cause changes to `tests/MyGame/Example/ArrayStruct.cs` ? Please make sure you run `generate_code.sh|bat` and include changes in this PR.
I generally don't think temporary allocations like this are going to be acceptable.

Reply:
I did `generate_code.sh` and it didn't affect the `ArrayStruct.cs` file. I'll delegate this to @mugisoba since he's more familiar with this.

Reply:
The `Create` method of a structure containing a fixed-size array requires arrays with all elements expanded as arguments.
We need new arrays to call this `Create` method.

### Example  TagList16 (above fbs)

```cs
public struct TagList16 : IFlatbufferObject
{
.
.
.
  public static Offset<test.TagList16> CreateTagList16(FlatBufferBuilder builder, int[] data_Id, float[] data_transform_rotation_X, float[] data_transform_rotation_Y, float[] data_transform_rotation_Z, float[] data_transform_rotation_W, float[] data_transform_translation_X, float[] data_transform_translation_Y, float[] data_transform_translation_Z) {
    builder.Prep(4, 512);
    for (int _idx0 = 16; _idx0 > 0; _idx0--) {
      builder.Prep(4, 32);
      builder.Prep(4, 28);
      builder.Prep(4, 12);
      builder.PutFloat(data_transform_translation_Z[_idx0-1]);
      builder.PutFloat(data_transform_translation_Y[_idx0-1]);
      builder.PutFloat(data_transform_translation_X[_idx0-1]);
      builder.Prep(4, 16);
      builder.PutFloat(data_transform_rotation_W[_idx0-1]);
      builder.PutFloat(data_transform_rotation_Z[_idx0-1]);
      builder.PutFloat(data_transform_rotation_Y[_idx0-1]);
      builder.PutFloat(data_transform_rotation_X[_idx0-1]);
      builder.PutInt(data_Id[_idx0-1]);
    }
    return new Offset<test.TagList16>(builder.Offset);
  }

// Object-API
  public static Offset<test.TagList16> Pack(FlatBufferBuilder builder, TagList16T _o) {
    if (_o == null) return default(Offset<test.TagList16>);
    var _data_id = new int[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_id[idx0] = _o.Data[idx0].Id;}
    var _data_transform_rotation_x = new float[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_transform_rotation_x[idx0] = _o.Data[idx0].Transform.Rotation.X;}
    var _data_transform_rotation_y = new float[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_transform_rotation_y[idx0] = _o.Data[idx0].Transform.Rotation.Y;}
    var _data_transform_rotation_z = new float[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_transform_rotation_z[idx0] = _o.Data[idx0].Transform.Rotation.Z;}
    var _data_transform_rotation_w = new float[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_transform_rotation_w[idx0] = _o.Data[idx0].Transform.Rotation.W;}
    var _data_transform_translation_x = new float[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_transform_translation_x[idx0] = _o.Data[idx0].Transform.Translation.X;}
    var _data_transform_translation_y = new float[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_transform_translation_y[idx0] = _o.Data[idx0].Transform.Translation.Y;}
    var _data_transform_translation_z = new float[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_transform_translation_z[idx0] = _o.Data[idx0].Transform.Translation.Z;}
    return CreateTagList16(
      builder,
      _data_id,
      _data_transform_rotation_x,
      _data_transform_rotation_y,
      _data_transform_rotation_z,
      _data_transform_rotation_w,
      _data_transform_translation_x,
      _data_transform_translation_y,
      _data_transform_translation_z);
  }
``

Reply:
And existing test cases will not be affected by this PR changes.

### before this PR.

Invalid code generated.
Nested struct array allocation and initialization were not being generated.

This pattern did not exist in test cases.

```cs
  public static Offset<test.TagList16> Pack(FlatBufferBuilder builder, TagList16T _o) {
    if (_o == null) return default(Offset<test.TagList16>);
    var _data_id = new int[16];
    for (var idx0 = 0; idx0 < 16; ++idx0) {_data_id[idx0] = _o.Data[idx0].Id;}
    return CreateTagList16(
      builder,
      _data_id,
      _data_transform_rotation_x,
      _data_transform_rotation_y,
      _data_transform_rotation_z,
      _data_transform_rotation_w,
      _data_transform_translation_x,
      _data_transform_translation_y,
      _data_transform_translation_z);
  }
```

Reply:
Ok, I appreciate hat this solves a bug, and maybe we should just merge it for that reason. But both the C++ generation code and the generated C# are not that great. In the case of the generated C#, we have all this data already stored, and then we needlessly transpose the data into arrays just to serialize it again. Use of `struct`, including nested ones, in because the user needs maximum efficiency, so use of lots of allocations and copies is particularly harmful. I'd appreciate a follow-up PR that changes the generated object API code such that we don't need these allocations and copies, either by special purposing the code and not calling `CreateTagList16`, or by introducing helper structs such that we don't need to transpose this data, maybe even changing how `CreateTagList16` is generated.